packages/core/src/types/overrides/overrides.tsx

```diff
@@ -0,0 +1,114 @@
+import type {
+  DeepPartial,
+  ExtractComponentNames,
+  RefineReactComponent,
+  RefineVueComponent,
+} from ".";
+
+export type RefineRuntime = "vue" | "react";
+
+export type RefineComponentName = ExtractComponentNames<
+  RefineOverrides<RefineRuntime>
+>;
+
+export type RefineComponent<
+  Name extends RefineComponentName,
+  Runtime extends RefineRuntime
+> = Runtime extends "vue"
+  ? RefineVueComponent<Name>
+  : Runtime extends "react"
+  ? RefineReactComponent<Name>
+  : never;
+
+/**
+ * Adds `class: string` for Vue and `className: string` for React.
+ */
+type ClassOrClassName<Runtime extends RefineRuntime> = Runtime extends "vue"
+  ? { class: string }
+  : Runtime extends "react"
+  ? { className: string }
+  : never;
+
+export type RefineOverrides<Runtime extends RefineRuntime> = {
+  group: ClassOrClassName<Runtime> & {
+    component: RefineComponent<"group", Runtime>;
+
+    addCriterionButton: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"group.addCriterionButton", Runtime>;
+
+      icon: ClassOrClassName<Runtime> & {
+        component: RefineComponent<"group.addCriterionButton.icon", Runtime>;
+      };
+    };
+  };
+
+  addGroupButton: ClassOrClassName<Runtime> & {
+    component: RefineComponent<"addGroupButton", Runtime>;
+  };
+
+  criterion: ClassOrClassName<Runtime> & {
+    component: RefineComponent<"criterion", Runtime>;
+
+    removeCriterionButton: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"criterion.removeCriterionButton", Runtime>;
+
+      icon: {
+        component: RefineComponent<
+          "criterion.removeCriterionButton.icon",
+          Runtime
+        >;
+      };
+    };
+  };
+
+  condition: ClassOrClassName<Runtime> & {
+    component: RefineComponent<"condition", Runtime>;
+  };
+
+  clause: ClassOrClassName<Runtime> & {
+    component: RefineComponent<"clause", Runtime>;
+  };
+
+  inputs: {
+    date: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"inputs.date", Runtime>;
+    };
+
+    doubleDate: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"inputs.doubleDate", Runtime>;
+    };
+
+    number: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"inputs.number", Runtime>;
+    };
+
+    doubleNumber: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"inputs.doubleNumber", Runtime>;
+    };
+
+    option: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"inputs.option", Runtime>;
+    };
+
+    relativeDate: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"inputs.relativeDate", Runtime>;
+    };
+
+    text: ClassOrClassName<Runtime> & {
+      component: RefineComponent<"inputs.text", Runtime>;
+    };
+  };
+
+  /**
+   * Used in many places within the query builder, e.g. selecting a condition, selecting a clause.
+   */
+  select: ClassOrClassName<Runtime> & {
+    component: RefineComponent<"select", Runtime>;
+  };
+};
```

Comment:
This is the core overrides type to be shared between Vue and React.

It looks a little funky, even for TypeScript as it is super generic. It uses a `ClassOrClassName<Runtime>` generic type to add either a `class` or `className` prop depending on if in Vue/React. It also connects to the component definitions for React or Vue, depending, in `packages/core/src/types/overrides/react.ts` and `packages/core/src/types/overrides/vue.ts`, respectively.

Reply:
Renamed to flavor so moving this comment over there.